tools/meteor-services/deploy.js

```diff
@@ -321,6 +323,206 @@ function canonicalizeSite(site) {
+        if (terminalBuildStatuses.indexOf(buildStatus) >= 0) {
+          if(buildStatus == 'success') {
+            Console.info('Successfully built app image, initiating deploy');
+            // Reset the polling rounds to poll faster for a change in deployStatus
+            // Basically, we don't necessarily expect deploying to take a long time simply
+            // because building did.
+            pollingRounds = 1;
+          } else {
+            // Any other terminal status is failure, so report on it
+            Console.info(`Your app failed to build, with buildStatus: ${buildStatus}`);
+            // We also want to stop polling, since deploy cannot succeed with a failed build
+            terminal = true;
+          }
+        } else if (buildStatus !== 'building') {
+          // This will report if the buildStatus changes to a non-terminal, unexpected state
+          //    e.g. failed-temporarily
+          Console.info(`Build state changed to nonterminal state: ${buildStatus}`)
+        }
+      } else if(currentStatus.deployStatus !== deployStatus) {
+        // Deploy status is treated similarly to buildStatus above
+        currentStatus.deployStatus = deployStatus;
+        if (terminalDeployStatuses.indexOf(deployStatus) >= 0) {
+          if(deployStatus == 'success') {
+            Console.info('Your app has been deployed successfully.' +
+              ' Waiting for new version to be active...');
+            // Reset the polling rounds to poll faster for a change in isActive
+            pollingRounds = 1;
+          } else {
+            Console.info(`Your app has failed to deploy, with deployStatus: ${deployStatus}`);
+            terminal = true;
+          }
+        } else if(deployStatus !== 'deploying') {
+          Console.info(`Deploy status changed to nonterminal state: ${deployStatus}`)
+        }
+      } else if(currentStatus.isActive !== isActive) {
+        currentStatus.isActive = isActive;
+        // isActive was initialized to false, so it can only be true
+        Console.info('Your deployed version is now active.');
+        terminal = true;
+      }
+    } else {
+      // If we did not get a valid Version Status response, just fail silently
+      // keep polling as per usual – this may have just been a whiff from Galaxy
+    }
+
+    let elapsed = new Date().getTime() - start;
+    // Poll again if version status isn't in a terminal state and we haven't exceeded the timeout
+    if(elapsed < timeoutMillis && !terminal) {
+      let newWaitTime = startIntervalMillis*Math.pow(backoffFactor, pollingRounds);
+
+      // Wait for exponentially increasingly more time, and then poll again
+      return new Promise(function (resolve) {
+        let timer = setTimeout(() => resolve(null), newWaitTime);
+      }).then(fiberHelpers.bindEnvironment(function() {
+        return pollForDeploy(pollingConfiguration, versionId, false);
+      })).await();
```

Comment:
This same pattern is repeated in other places—let's hoist this into a reusable component?

Reply:
it already is, utils.sleepMs.